### PR TITLE
Add account pending limit and lifetime in TxPool

### LINKED ISSUE #226: Txpool pending and queued sets are never pruned, causing eventual overflow or DOS
### Description

There is no mechanism to reduce transactions that might be eventually valid from the pending or queued sets, causing perpetual growth. 

### Your environment

* OS and version Ubuntu 20
* version of the Polygon SDK 6efebc6a92f885082620ed2a8d59e37d4d27074c
* branch that causes this issue `develop`

### Steps to reproduce

* Tell us how to reproduce this issue <br />
Create a chain with 4 validators. Create an a premine with a balance of 100 ETH. Submit txns with nonce 0...1000 and submit **them immediately without waiting for receipt**, each sending 1 ETH to some random account. Observe the first ~100 transactions get mined. Now the txn pool `pendingQueue` should contain 900 transactions.

These 900 transactions will constantly be popped then re-pushed back onto the queue because they are thrown with a recoverable error which pushes them back onto the queue

https://github.com/0xPolygon/polygon-sdk/blob/6efebc6a92f885082620ed2a8d59e37d4d27074c/consensus/ibft/ibft.go#L406

```
2021-11-18T16:24:28.038-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
```

* Where the issue is, if you know <br />

* Which commands triggered the issue, if any

`./main server`

### Expected behaviour

Transactions that are invalid for a long time should eventually be dropped. 

### Actual behaviour

Transactions that are invalid are constantly re-pushed onto the queue as most of the errors thrown here are recoverable

https://github.com/0xPolygon/polygon-sdk/blob/6efebc6a92f885082620ed2a8d59e37d4d27074c/state/executor.go#L358

### Logs

`txpool_status` on our testnet validator nodes

```
{ jsonrpc: '2.0', id: null, result: { pending: 0, queued: 1 } }
{ jsonrpc: '2.0', id: null, result: { pending: 0, queued: 42 } }
{ jsonrpc: '2.0', id: null, result: { pending: 0, queued: 329 } }
{ jsonrpc: '2.0', id: null, result: { pending: 3292, queued: 1561 } }
{ jsonrpc: '2.0', id: null, result: { pending: 3968, queued: 22 } }
{ jsonrpc: '2.0', id: null, result: { pending: 1145, queued: 14379 } }
{ jsonrpc: '2.0', id: null, result: { pending: 69, queued: 5711 } }
{ jsonrpc: '2.0', id: null, result: { pending: 109, queued: 5629 } }
{ jsonrpc: '2.0', id: null, result: { pending: 256, queued: 5559 } }
{ jsonrpc: '2.0', id: null, result: { pending: 0, queued: 42 } }
{ jsonrpc: '2.0', id: null, result: { pending: 60, queued: 5736 } }
{ jsonrpc: '2.0', id: null, result: { pending: 3706, queued: 1641 } }
{ jsonrpc: '2.0', id: null, result: { pending: 3630, queued: 1389 } }
{ jsonrpc: '2.0', id: null, result: { pending: 0, queued: 204 } }
```

```
2021-11-18T16:23:40.026-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:23:48.026-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:23:56.022-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:04.025-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:12.027-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:20.027-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:20.027-0500 [INFO]  polygon.consensus.ibft: picked out txns from pool: num=0 remaining=985
2021-11-18T16:24:28.038-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:28.038-0500 [INFO]  polygon.consensus.ibft: picked out txns from pool: num=0 remaining=985
2021-11-18T16:24:36.024-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:36.024-0500 [INFO]  polygon.consensus.ibft: picked out txns from pool: num=0 remaining=985
2021-11-18T16:24:44.028-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:24:52.026-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:25:00.025-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:25:00.025-0500 [INFO]  polygon.consensus.ibft: picked out txns from pool: num=0 remaining=985
2021-11-18T16:25:08.024-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
2021-11-18T16:25:08.024-0500 [INFO]  polygon.consensus.ibft: picked out txns from pool: num=0 remaining=985
2021-11-18T16:25:16.023-0500 [ERROR] polygon: failed to apply tx: err="not enough funds to cover gas costs, recoverable [true]"
```

Please paste any logs here that demonstrate the issue, if they exist

### Proposed solution

There should be some sort of periodic pool pruning as in geth. Geth has a periodic mechanism:

https://github.com/maticnetwork/bor/blob/a9c71eff578c652ddfe00a73267454bed5957ed0/core/tx_pool.go#L1137

And the algorithms:

https://github.com/maticnetwork/bor/blob/a9c71eff578c652ddfe00a73267454bed5957ed0/core/tx_pool.go#L1377
https://github.com/maticnetwork/bor/blob/a9c71eff578c652ddfe00a73267454bed5957ed0/core/tx_pool.go#L1462

### PR DESCRIPTION
**Pull request by Kourin1996**
# Description

Relevant: https://github.com/0xPolygon/polygon-sdk/issues/226
PR for Docs: https://github.com/0xPolygon/polygon-sdk-docs/pull/30

This PR adds account limit in pending queue and truncating by lifetime in order to protect from dos attack.

In https://github.com/0xPolygon/polygon-sdk/issues/226 is reporting the fact there would be many stuck transactions if the account sends many transactions even though it doesn't have enough funds. TxPool check account's balance to validate incoming transactions but they can pass if the account sends transactions before confirmations of previous transactions.

This PR adds 2 feature against this issue.
* AccountLimit is the limit of num transactions in pending queue. AccountGauge is counting the num txs in pending queue and TxPool can promote up to AccountLimit-numPending.
* Lifetime is maximum amount of time non-executable transaction are queued. Account's timestamp is updated when new Tx is added to TxPool and try to find inactive account every 1 mins. Once TxPool founds inactive account, try to remove its transactions from account queue.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [x] I have assigned this PR to myself
- [ ] I have added at least 1 reviewer
- [x] I have tested this code
- [x] I have updated the README and other relevant documents (guides...)
- [ ] I have added sufficient documentation both in code, as well as in the READMEs

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: txpool/txpool.go [hunk 1]
```diff
@@ -112,41 +141,51 @@ type TxPool struct {
...
+		sealing:             config.Sealing,
+		locals:              newLocalAccounts(config.Locals),
+		noLocals:            config.NoLocals,
+		priceLimit:          config.PriceLimit,
+		accountPendingLimit: config.AccountPendingLimit,
+		lifetime:            config.Lifetime,
+		forks:               forks,
+		closeCh:             make(chan struct{}, 1),
 	}
+	fmt.Printf("accountPendingLimit=%d, lifetime=%+v\n", txPool.accountPendingLimit, txPool.lifetime)
```
Comment (brkomir) [inline txpool/txpool.go]: is this a leftover?

File: command/dev/dev_command.go [hunk 2]
```diff
@@ -72,6 +72,22 @@ func (d *DevCommand) DefineFlags() {
...
+	d.FlagMap["account-pending-limit"] = helper.FlagDescriptor{
+		Description: fmt.Sprintf("Sets maximum number of transactions in pending queue per account. Default: %d", helper.DefaultConfig().TxPool.AccountPendingLimit),
+		Arguments: []string{
+			"ACCOUNT_PENDING_LIMIT",
+		},
+		FlagOptional: true,
+	}
+
+	d.FlagMap["lifetime"] = helper.FlagDescriptor{
+		Description: fmt.Sprintf("Maximum amount of time non-executable transaction are queued. Default: %s", helper.DefaultConfig().TxPool.Lifetime),
```
Comment (brkomir) [inline command/dev/dev_command.go]: Just a small correction in the message, i suggest 
"Sets the maximum amount of time a non-executable transaction can wait in the queue. Default: %s"

File: command/server/server_command.go [hunk 3]
```diff
@@ -174,6 +174,22 @@ func (c *ServerCommand) DefineFlags() {
...
+	c.flagMap["account-pending-limit"] = helper.FlagDescriptor{
+		Description: fmt.Sprintf("Sets maximum number of transactions in pending queue per account. Default: %d", helper.DefaultConfig().TxPool.AccountPendingLimit),
+		Arguments: []string{
+			"ACCOUNT_PENDING_LIMIT",
+		},
+		FlagOptional: true,
+	}
+
+	c.flagMap["lifetime"] = helper.FlagDescriptor{
+		Description: fmt.Sprintf("Sets maximum amount of time transactions are queued in account queue. Default: %s", helper.DefaultConfig().TxPool.Lifetime),
```
Comment (brkomir) [inline command/server/server_command.go]: Similar as the comment in dev_command.go

**Review by brkomir — APPROVED** [review]
LGTM

File: txpool/txpool.go (hunk 1 above)
Comment (Kourin1996) [reply txpool/txpool.go]: Thank you for finding! Removed at https://github.com/0xPolygon/polygon-sdk/pull/239/commits/dbd69ad363159608b6b616275f93dcda242b29bf

File: command/dev/dev_command.go (hunk 2 above)
Comment (Kourin1996) [reply command/dev/dev_command.go]: That's better :)
Fixed at https://github.com/0xPolygon/polygon-sdk/pull/239/commits/e97a6a45c28c3d68b0a8b3efab0a2ac775fe598d

File: command/server/server_command.go (hunk 3 above)
Comment (Kourin1996) [reply command/server/server_command.go]: Fixed at https://github.com/0xPolygon/polygon-sdk/pull/239/commits/e97a6a45c28c3d68b0a8b3efab0a2ac775fe598d

**Review by zivkovicmilos — APPROVED** [review]
I missed reviewing @Kourin1996's PRs, always learn something new 👍 

Looks great 🚀

**Review by zivkovicmilos — APPROVED** [review]
Thank you for fixing these minor issues 🙏 

Approving again 💯

**Comment by zivkovicmilos** [comment]
Closing and archiving under `archive/feature/prune-txs-in-txpool`

https://github.com/0xPolygon/polygon-sdk/releases/tag/archive%2Ffeature%2Fprune-txs-in-txpool

**Comment by zivkovicmilos** [comment]
Closing and archiving under `archive/feature/prune-txs-in-txpool`

https://github.com/0xPolygon/polygon-sdk/releases/tag/archive%2Ffeature%2Fprune-txs-in-txpool

**Pull request closed without merging**